data/transactions/logic/eval.go

```diff
@@ -451,42 +456,82 @@ const MaxStackDepth = 1000
 
 func (cx *evalContext) step() {
 	opcode := cx.program[cx.pc]
-	if opsByOpcode[opcode].op == nil {
-		cx.err = fmt.Errorf("%3d illegal opcode %02x", cx.pc, opcode)
+	spec := opsByOpcode[cx.version][opcode]
+	if spec.op == nil {
+		cx.err = fmt.Errorf("%3d illegal opcode 0x%02x", cx.pc, opcode)
+		return
+	}
+	if (cx.runModeFlags & spec.Modes) == 0 {
+		cx.err = fmt.Errorf("%s not allowed in current mode", spec.Name)
 		return
 	}
-	argsTypes := opsByOpcode[opcode].Args
+	if spec.Version > cx.version {
+		cx.err = fmt.Errorf("%s not allowed in program version %d", spec.Name, cx.version)
+		return
+	}
+	argsTypes := spec.Args
 	if len(argsTypes) >= 0 {
```

Comment:
isn't this always true?

Reply:
it is, @algobolson do you remember the rationale?